src/Io/RequestHeaderParser.php

```diff
@@ -124,32 +124,34 @@ public function parseRequest($headers, $remoteSocketUri, $localSocketUri)
     {
         // additional, stricter safe-guard for request line
         // because request parser doesn't properly cope with invalid ones
-        if (!\preg_match('#^[^ ]+ [^ ]+ HTTP/\d\.\d#m', $headers)) {
+        $start = array();
+        if (!\preg_match('#^(?<method>[^ ]+) (?<target>[^ ]+) HTTP/(?<version>\d\.\d)#m', $headers, $start)) {
             throw new \InvalidArgumentException('Unable to parse invalid request-line');
         }
 
-        // parser does not support asterisk-form and authority-form
-        // remember original target and temporarily replace and re-apply below
-        $originalTarget = null;
-        if (\strncmp($headers, 'OPTIONS * ', 10) === 0) {
-            $originalTarget = '*';
-            $headers = 'OPTIONS / ' . \substr($headers, 10);
-        } elseif (\strncmp($headers, 'CONNECT ', 8) === 0) {
-            $parts = \explode(' ', $headers, 3);
-            $uri = \parse_url('tcp://' . $parts[1]);
+        // only support HTTP/1.1 and HTTP/1.0 requests
+        if ($start['version'] !== '1.1' && $start['version'] !== '1.0') {
+            throw new \InvalidArgumentException('Received request with invalid protocol version', 505);
+        }
 
-            // check this is a valid authority-form request-target (host:port)
-            if (isset($uri['scheme'], $uri['host'], $uri['port']) && count($uri) === 3) {
-                $originalTarget = $parts[1];
-                $parts[1] = 'http://' . $parts[1] . '/';
-                $headers = implode(' ', $parts);
-            } else {
-                throw new \InvalidArgumentException('CONNECT method MUST use authority-form request target');
-            }
+        $matches = array();
+        $n = \preg_match_all('/^([^()<>@,;:\\\"\/\[\]?={}\x01-\x20\x7F]++):[\x20\x09]*+((?:[\x20\x09]*+[\x21-\x7E\x80-\xFF]++)*+)[\x20\x09]*+/m', $headers, $matches, \PREG_SET_ORDER);
+
+        if (\substr_count($headers, "\n") !== $n) {
+            throw new \InvalidArgumentException('Unable to parse invalid request header fields');
         }
 
-        // parse request headers into obj implementing RequestInterface
-        $request = g7\parse_request($headers);
```

Comment:
You can now also drop the `use RingCentral\Psr7 as g7;` at the top

Reply:
Good catch, updated :shipit: